### ISSUE TITLE
Charts may not be fully deleted

### ISSUE THREAD

**Issue by Rmelikyan** [pyvista/pyvista#5338]
### Describe the bug, what's wrong, and what you expected.

tl;dr: Calling remove chart does not actually delete the chart and all references to it. This led to unintended behavior where if a new chart was created and placed on top of old (not fully deleted chart) then the new chart would default to bottom lefthand position rather than input position. Attached code shows minimum working example of behavior.

I have been working on a pyvista app which allows a user to flick through time series point-cloud data using key bindings. I've been progressively upgrading the program to include more functionality such as changing scalar-data and colormap with button widgets and most recently displaying an overlaying chart that should update along with the updating point cloud. My naive first approach was, at each time step (call of `plotter.update()`) I would generate the figure of interest with matplotlib, delete the old pyvista chart and create and add a new chart with `pyvista.ChartMPL` and `plotter.add_chart`. Oddly, the first plot always looked great, but the second would never show up in the specified location, instead would default to the bottom lefthand corner. I got the suspicion that maybe the old chart wasn't being deleted properly (`del(chart)` throws errors and `plotter.remove_chart()` didn't do the trick). So I started digging into this and found that even after calling the remove_chart function, that the renderer still thought it had a chart (`renderer.has_charts` gave True).

Looking into the source, I could find no way to actually entirely delete a chart from a plotter instance.

Luckily, I realized that the chart `redraw_on_render` attribute, which I had turned off for increased performance when the chart was not visible (my chart button is also the toggle), was actually the solution I was looking for. So now I just update the matplotlib figure on each render which updates the chart.

That being said, I still wanted to provide the below example and inquire if this is the intended behavior.

Thanks!

### Steps to reproduce the bug.

```
import pyvista as pv
pl = pv.Plotter()
assert not pl.renderer.has_charts
chart = pv.Chart2D()
chart.plot(range(10), range(10))
pl.add_chart(chart)
assert pl.renderer.has_charts
print("A chart correctly exists and can be accessed")
print(pl.renderer._charts)
print(pl.renderer._charts[0])
pl.remove_chart(chart)
try:
    assert not pl.renderer.has_charts
except Exception as error:
    print(type(error).__name__)
    print(f"Called pl.remove_chart but still renderer.has_charts == {pl.renderer.has_charts}")
    print(pl.renderer._charts)
print("But the chart cannot be accessed due to index error")
try:
    print(pl.renderer._charts[0])
except Exception as error:
    print(type(error).__name__)
```

### System Information

```shell
--------------------------------------------------------------------------------
  Date: Wed Dec 13 20:47:10 2023 MST

                OS : Darwin
            CPU(s) : 16
           Machine : x86_64
      Architecture : 64bit
               RAM : 64.0 GiB
       Environment : Python
       File system : apfs
        GPU Vendor : ATI Technologies Inc.
      GPU Renderer : AMD Radeon Pro 5500M OpenGL Engine
       GPU Version : 4.1 ATI-4.6.21
  MathText Support : False

  Python 3.9.17 (main, Jun 20 2023, 17:03:39)  [Clang 13.0.0
  (clang-1300.0.29.30)]

           pyvista : 0.42.3
               vtk : 9.2.6
             numpy : 1.23.3
        matplotlib : 3.7.2
            scooby : 0.7.4
             pooch : v1.6.0
            pillow : 9.2.0
           IPython : 8.5.0
        ipywidgets : 8.1.1
             scipy : 1.9.1
              tqdm : 4.64.1
             trame : 3.2.7
      trame_client : 2.12.5
      trame_server : 2.12.0
         trame_vtk : 2.5.9
     trame_vuetify : 2.3.1
      nest_asyncio : 1.5.6
--------------------------------------------------------------------------------
```

**Comment by dcbr** [comment]
So upon closer inspection this is actually intended behaviour and not a bug. The renderer's `has_charts` property is only used for internal bookkeeping (and should thus be made a private property) to make sure the associated context scene and actor object are properly cleaned up. It only means that a `Charts` object (collecting all individual charts) exists for that renderer (so at some point in the renderer's existence there was a chart added to it), not that there are any actual charts in there at that moment. It's for sure a confusing name, so that's why it's best changed in my opinion.

The other things you described in your post are more concerning to me.

> Oddly, the first plot always looked great, but the second would never show up in the specified location, instead would default to the bottom lefthand corner. I got the suspicion that maybe the old chart wasn't being deleted properly (`del(chart)` throws errors and `plotter.remove_chart()` didn't do the trick).

Do you have some example code for that, because that shouldn't happen?

**Comment by Rmelikyan** [comment]
Wow, I don't know what I did before but I can't reproduce the bug... Unfortunately I didn't push any version control during the bug so I only have a before and after that are both working (though the before had the problem of creating many mpl figures where now I'm clearing and re-using the same figure). I'm providing a working example of what I was trying to do that, for unknown reasons, is working now but not before. Maybe somewhere in the additional complications of my project something else was causing mischief...

```import pyvista as pv
import numpy as np
import matplotlib.pyplot as plt

def update_chart():
    if pl.renderer.has_charts:
        pl.remove_chart(-1)
    fig, ax = plt.subplots()
    ax.plot(range(10), np.random.random(10), lw=2, c='r')
    chart = pv.ChartMPL(fig, size=(0.46, 0.25), loc=(0.52, 0.7), redraw_on_render=False)
    pl.add_chart(chart)
def update_figure():
    update_chart()
    cloud = np.random.random((1000, 3))
    pl.add_points(cloud, render=False, name='cloud')
    pl.update()

pl = pv.Plotter(notebook=False)
cloud = np.random.random((1000, 3))
pl.add_points(cloud, render=False, name='cloud')
pl.add_key_event('u', update_figure)
pl.show(auto_close=False, interactive=True, interactive_update=True)
pl.show()
```

**Comment by dcbr** [comment]
Alright, in case you still encounter any problems later, let me know!
I modified the internal bookkeeping in #5356 and kept `has_charts` as a public property that now properly returns `True` as you would have expected (i.e. only `True` when there are charts in the collection).
